lib/core/include/irods/configuration_hook_manager.hpp

```diff
@@ -0,0 +1,117 @@
+    /// the hooks associated with
+    /// "/advanced_settings/delay_server_sleep_time_in_seconds", also in order
+    /// of addition.
+    using hook = std::function<void(const std::string&, const nlohmann::json&)>;
+
+    namespace
+    {
+        // This is where the function objects are stored, given that they will not be removed or reordered, their index
+        // is what edge_sets' value refers to.
+        std::vector<hook> hooks;
+
+        // Edge set, as in edge-triggered, contains an association between the changed property and an index into hooks.
+        std::unordered_multimap<std::string, size_t> edge_sets;
+    } // namespace
+
+    /// \brief Adds a hook to be run whenever a key on the list of hooks changes.
+    ///        \see hook
+    /// \param [in] hook_triggers An array or container of json paths to match, this
+    ///             container must support begin and end iterators.
+    ///
+    ///             The special value CATCHALL_KEYWORD matches any property that has changed and gets run first
+    /// \param [in] hook The function to run when the json path matches.
+    ///
+    /// \b Example
+    /// \code{.cpp}
+    /// namespace cr = irods::experimental::configuration_reload;
+    /// cr::add_hook({cr::CATCHALL_KEYWORD}, [&](const std::string& json_path,
+    ///                                          const nlohmann::json& json_patch_op)
+    /// {
+    ///    if(json_path.substr(0,29) != "/example_configuration_object") {
+    ///         return;
+    ///    }
+    ///    handle_update(json_patch_op);
+    /// });
+    /// cr::add_hook({"/example_configuration_object/circumference_to_diameter_ratio"},
+    ///              [](const std::string& json_path, const nlohmann::json& json_patch_op){
+    ///    // Update the "PI" global value in the static lua state.
+    ///    if(json_patch_op.at("op")=="insert"||json_patch_op.at("op")=="replace"){
+    ///        lua_pushnumber( Lua_state, json_patch_op.at("value").get<double>());
+    ///        lua_setglobal( Lua_state, "PI");
+    ///    }
+    /// });
+    /// \endcode
+    void add_hook(const auto& hook_triggers, hook&& hook)
+    {
+        size_t id = hooks.size();
+        // Take ownership, we don't need to deal with any missing closures.
+        // those can be nasty to debug.
+        hooks.emplace_back(std::move(hook));
+        for (const auto& name : hook_triggers) {
+            edge_sets.insert(std::make_pair(name, id));
+        }
+    }
+
+    /// \brief Run the hooks, the changes are a list of json patches.
+    /// \param [in] changes The list of changes in json patch format, see
+    /// RFC6902 for more details about that format.
+    /// \code{.cpp}
+    /// namespace cr = irods::experimental::configuration_reload;
+    /// cr::add_hook({cr::CATCHALL_KEYWORD}[](const std::string& json_path, const nlohmann::json& patch){
+    ///    std::cout<<patch<<std::endl;
+    /// });
```

Comment:
Missing call to `run_hooks()`.

Reply:
I mean, it's not *really* meant to be run by the user normally :p 

I should mention that somewhere shouldn't I?

Reply:
Absolutely, but consider developers who are writing clients. They may want to use this. For those cases, they would have to work out when `run_hooks()` should be invoked in their code.

Or perhaps I'm misunderstanding something?

Reply:
Fair enough, I hadn't considered there being legitimate uses for it on the client side.

Reply:
Also putting it in the examples makes them way easier to write